File: day02/julia/part2.jl
#!/usr/bin/env julia

using FromFile

@from "Helper.jl" import Helper

function main()
    # fname = "example.txt"
    fname = "input.txt"
    lines = Helper.read_lines(fname)

    horizontal, depth, aim = (0, 0, 0)

    for line in lines
        parts = split(line)
        instruction = parts[1]
        value = parse(Int, parts[2])
        if instruction == "forward"
            horizontal += value
            depth += aim * value
        elseif instruction == "down"
            aim += value
        elseif instruction == "up"
            aim -= value
        else
            @assert false
        end
    end
    println(horizontal * depth)
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: day02/python/part2.py
#!/usr/bin/env python3

import helper

class ShouldNeverGetHere(Exception):
    pass

def main():
    # lines = helper.read_lines("example.txt")
    lines = helper.read_lines("input.txt")

    horizontal, depth, aim = (0, 0, 0)

    for line in lines:
        parts = line.split()
        instruction = parts[0]
        value = int(parts[1])
        if instruction == "forward":
            horizontal += value
            depth += aim * value
        elif instruction == "down":
            aim += value
        elif instruction == "up":
            aim -= value
        else:
            raise ShouldNeverGetHere()
        #
    #
    print(f"{horizontal=}")
    print(f"{depth=}")
    print(f"{aim=}")
    print("---")
    print(horizontal * depth)

##############################################################################

if __name__ == "__main__":
    main()
